starlite/datastructures/provide.py

```diff
@@ -1,27 +1,79 @@
+from inspect import Parameter, Signature
 from typing import TYPE_CHECKING, Any, Optional, cast
 
+from anyio import Lock
+
 from starlite.types import Empty
 from starlite.utils.sync import AsyncCallable, is_async_callable
 
 if TYPE_CHECKING:
     from typing import Type
 
+    from starlite.connection import Request
     from starlite.signature import SignatureModel
     from starlite.types import AnyCallable
 
 
+def _wrap_cache_for_request(fn: "AnyCallable", override_name: Optional[str]) -> "AnyCallable":
+    """Wrap the function to cache the result per request."""
+    if override_name:
+        key = override_name
+    else:
+        key = fn.__name__
+    lock = Lock()
+    wrapped_sig = sig = Signature.from_callable(fn)
+    include_request = True
+    if "request" not in sig.parameters:
+        include_request = False
+        params = list(sig.parameters.values())
+        params.append(Parameter("request", Parameter.KEYWORD_ONLY, annotation="Request"))
```

Comment:
clever. 

I would though not do this here - instead of manipulating the signature model to add this parameter, I would add resolution logic to handle `cache_per_request` the kwargs logic: https://github.com/starlite-api/starlite/blob/main/starlite/kwargs.py

Reply:
I moved the logic over to the KwargsModel, is this what you meant?